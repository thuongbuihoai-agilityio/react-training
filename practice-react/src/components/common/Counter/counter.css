.counter {
  position: relative;
  z-index: 15;
  width: 205px;
}

.counter-input {
  border: 1px solid var(--border-primary);
  border-radius: 24.5px;
  text-align: center;
  padding: 14px;
}

.counter-minus {
  position: absolute;
  z-index: 20;
  top: 33%;
  left: 22px;
}

.counter-plus {
  position: absolute;
  z-index: 20;
  top: 35%;
  right: 25px;
}