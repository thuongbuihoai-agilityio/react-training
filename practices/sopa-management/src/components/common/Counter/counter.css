.counter {
  position: relative;
  border-radius: var(--rounded-secondary);
  border: 1px solid var(--border-primary);
  height: var(--h-normal);
  width: var(--w-normal);
}

.counter-minus,
.counter-minus-disable {
  height: 44px;
  position: absolute;
  z-index: 2;
  left: 20px;
}

.counter-minus-disable {
  cursor: not-allowed;
}

.counter-plus {
  position: absolute;
  top: 22px;
  left: 121px;
}

.counter-minus:before,
.counter-minus-disable:before,
.counter-plus:before,
.counter-plus:after {
  content: '';
  position: absolute;
  background-color: var(--text-tertiary);
  height: 2px;
  width: 10px;
  top: 50%;
  left: 50%;
  margin-top: -1px;
  margin-left: -5px;
}

.counter-plus:after {
  transform: rotate(90deg);
}

.counter-input {
  position: absolute;
  font-size: var(--text-base);
  font-weight: var(--font-semi-bold);
  border: none;
  width: 50px;
  left: 34%;
  top: 10px;
  text-align: center;
}

.counter-input::-webkit-outer-spin-button,
.counter-input::-webkit-inner-spin-button {
  -webkit-appearance: none;
  margin: 0;
}
