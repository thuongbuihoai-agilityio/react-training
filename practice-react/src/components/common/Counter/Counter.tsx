import React, { memo, useCallback, useState } from "react";
import Icon from "../Icon/Icon";
import "./counter.css";

interface CounterProps {
  counter?: number;
}

const Counter: React.FC<CounterProps> = memo(({ counter }) => {
  const [count, setCount] = useState<number>(counter!);
  const [quantity, setQuantity] = useState<{}>("")

  const handleIncrease = useCallback(() => {
    setCount(count + 1);
  }, [count]);

  const handleDecrease = useCallback(() => {
    setCount(count - 1);
  }, [count]);

  // handle change value
  const handleChange = useCallback(
    (event: { target: HTMLInputElement }) => {
      const value = event.target.value;
      const fieldName = event.target.name;
      setQuantity({...quantity, [fieldName]: value})
    },
    [count]
  );

  return (
    <div data-testid="counter" className="counter">
      <div className="counter-minus">
        <Icon onClick={handleDecrease} iconName="minus" />
      </div>
      <input
        className="counter-input"
        name="productQuantity"
        min={0}
        value={count}
        onChange={handleChange}
      />
      <div className="counter-plus">
        <Icon onClick={handleIncrease} iconName="plus" />
      </div>
    </div>
  );
});

export default Counter;
