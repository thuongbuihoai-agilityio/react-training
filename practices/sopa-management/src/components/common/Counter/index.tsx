import { memo } from 'react';

// Components
import {
  Button,
  ButtonType,
  Input
} from '@components/common';

// Styles
import './counter.css';

interface CounterProps {
  value: number;
  onIncrement: () => void;
  onDecrement: () => void;
};

const Counter = ({
  value,
  onIncrement,
  onDecrement
}: CounterProps) => {
  const isDisable = value === 1;

  return (
    <div data-testid='counter' className='counter'>
      <Button
        ariaLabel='Decrement'
        type={ButtonType.btnIconPrimary}
        className={isDisable ? 'counter-minus-disable' : 'counter-minus'}
        disable={isDisable}
        onClick={onDecrement}
      />
      <Input type='number' classNameInput='counter-input' value={value} />
      <Button
        ariaLabel='Increment'
        type={ButtonType.btnIconPrimary}
        className='counter-plus'
        onClick={onIncrement}
      />
    </div>
  );
}

export default memo(Counter);
