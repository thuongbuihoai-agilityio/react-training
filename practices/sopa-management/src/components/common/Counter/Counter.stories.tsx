// Libs
import type { Meta, StoryObj } from '@storybook/react';

// Components
import Counter from '@common/Counter';

const meta: Meta<typeof Counter> = {
  component: Counter,
};

export default meta;
type Story = StoryObj<typeof Counter>;

export const CounterDefault: Story = {
  args: {
    value: 1,
  }
};
