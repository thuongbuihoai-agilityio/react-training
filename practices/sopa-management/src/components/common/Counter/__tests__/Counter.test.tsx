// Libs
import '@testing-library/jest-dom';
import { render } from '@testing-library/react';

// Components
import Counter from '@common/Counter';

describe('Counter component', () => {
  const props = {
    value: 2,
    onIncrement: jest.fn(),
    onDecrement: jest.fn()
  };
  test('should render Counter component', () => {
    const { getByTestId } = render(<Counter {...props} />);
    expect(getByTestId('counter')).toBeInTheDocument();
  });

  test('matches snapshot', () => {
    const { asFragment } = render(<Counter {...props} />);
    expect(asFragment()).toMatchSnapshot();
  });
});
