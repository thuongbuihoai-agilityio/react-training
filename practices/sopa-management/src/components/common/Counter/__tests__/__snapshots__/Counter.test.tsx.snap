// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Counter component matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="counter"
    data-testid="counter"
  >
    <button
      aria-label="Decrement"
      class="counter-minus btn btn-icon-primary"
      data-testid="button"
    />
    <input
      aria-hidden="true"
      class=" counter-input input-"
      data-testid="input"
      id=""
      name=""
      type="number"
      value="2"
    />
    <button
      aria-label="Increment"
      class="counter-plus btn btn-icon-primary"
      data-testid="button"
    />
  </div>
</DocumentFragment>
`;
